src/passes/TypeGeneralizing.cpp

```diff
@@ -0,0 +1,438 @@
+
+  TransferFn(Module& wasm, Function* func)
+    : wasm(wasm), func(func), lattice(func) {}
+
+  Type pop() noexcept { return lattice.pop(*state); }
+  void push(Type type) noexcept { lattice.push(*state, type); }
+  void clearStack() noexcept { lattice.clearStack(*state); }
+  Type getLocal(Index i) noexcept { return lattice.getLocal(*state, i); }
+  void updateLocal(Index i, Type type) noexcept {
+    // TODO: Collect possible successor blocks that might depend on an updated
+    // local requirement here.
+    return lattice.updateLocal(*state, i, type);
+  }
+
+  void dumpState() {
+#if TYPE_GENERALIZING_DEBUG
+    std::cerr << "locals: ";
+    for (size_t i = 0, n = lattice.getLocals(*state).size(); i < n; ++i) {
+      if (i != 0) {
+        std::cerr << ", ";
+      }
+      std::cerr << getLocal(i);
+    }
+    std::cerr << "\nstack: ";
+    auto& stack = std::get<1>(*state);
+    for (size_t i = 0, n = stack.size(); i < n; ++i) {
+      if (i != 0) {
+        std::cerr << ", ";
+      }
+      std::cerr << stack[i];
+    }
+    std::cerr << "\n";
+#endif // TYPE_GENERALIZING_DEBUG
+  }
+
+  std::vector<const BasicBlock*>
+  transfer(const BasicBlock& bb, typename State::Element& elem) noexcept {
+    DBG(std::cerr << "transferring bb " << bb.getIndex() << "\n");
+    state = &elem;
+    // This is a backward analysis: The constraints on a type depend on how it
+    // will be used in the future. Traverse the basic block in reverse.
+    dumpState();
+    if (bb.isExit()) {
+      DBG(std::cerr << "visiting exit\n");
+      visitFunctionExit();
+      dumpState();
+    }
+    for (auto it = bb.rbegin(); it != bb.rend(); ++it) {
+      DBG(std::cerr << "visiting " << ShallowExpression{*it} << "\n");
+      visit(*it);
+      dumpState();
+    }
+    DBG(std::cerr << "\n");
+
+    state = nullptr;
+
+    // Return the blocks that may need to be re-analyzed.
+    const auto& preds = bb.preds();
+    std::vector<const BasicBlock*> dependents;
+    dependents.reserve(preds.size());
+    dependents.insert(dependents.end(), preds.begin(), preds.end());
+    return dependents;
+  }
+
+  void visitFunctionExit() {
+    // We cannot change the types of parameters, so require that they have their
+    // original types.
+    Index i = 0;
+    Index numParams = func->getNumParams();
+    Index numLocals = func->getNumLocals();
+    for (; i < numParams; ++i) {
+      updateLocal(i, func->getLocalType(i));
+    }
+    // We also cannot change the types of any other non-ref locals. For
+    // reference-typed locals, we cannot generalize beyond their top type.
+    for (; i < numLocals; ++i) {
+      auto type = func->getLocalType(i);
+      // TODO: Support optimizing tuple locals.
+      if (type.isRef()) {
+        updateLocal(i, Type(type.getHeapType().getTop(), Nullable));
+      } else {
+        updateLocal(i, type);
+      }
+    }
+    // We similarly cannot change the types of results. Push requirements that
```

Comment:
The pushed requirement is that the stack end up with the correct type. I'll make "requirements" singular here since there is only one. (At one point during development this was a loop, so there were multiple requirements.)